main.go

```diff
@@ -108,15 +112,55 @@ func init() {
 	prometheus.MustRegister(version.NewCollector("blackbox_exporter"))
 }
 
+func probeRequestHandler(w http.ResponseWriter, r *http.Request) {
+	sc.RLock()
+	c := sc.C
+	sc.RUnlock()
+
+	// If a timeout is configured via the Prometheus header, add it to the request.
+	var prometheusTimeout string
+	if r.Header["X-Prometheus-Scrape-Timeout-Seconds"] != nil {
+		prometheusTimeout = r.Header["X-Prometheus-Scrape-Timeout-Seconds"][0]
+	}
+
+	timeoutSeconds, err := strconv.ParseFloat(prometheusTimeout, 64)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to parse timeout from Prometheus header: %s", err), http.StatusInternalServerError)
+	}
+	if timeoutSeconds == 0 {
+		timeoutSeconds = 10
+	}
+
+	moduleName := r.URL.Query().Get("module")
+	if moduleName == "" {
+		moduleName = "http_2xx"
+	}
+	module, ok := c.Modules[moduleName]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
+	}
+
+	if module.Timeout.Seconds() < timeoutSeconds && module.Timeout.Seconds() > 0 {
+		timeoutSeconds = module.Timeout.Seconds()
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
+	defer cancel()
+	r = r.WithContext(ctx)
+
+	probeHandler(ctx, w, r, c)
+}
+
+var timeoutOffset = flag.Float64("timeout-offset", 0.5, "Offset to subtract from timeout in seconds.")
```

Comment:
Keep all the flags together, they should be up near the top of this file in a var section.